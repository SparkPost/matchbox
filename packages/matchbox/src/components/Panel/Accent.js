import React from 'react';
import { Box } from '../Box';

function Accent(props) {
  const color = React.useMemo(() => {
    switch (props.accentColor) {
      case 'orange':
        return 'brand.orange';
      case 'green':
        return 'green.700';
      case 'yellow':
        return 'yellow.400';
      case 'red':
        return 'red.700';
      case 'gray':
        return 'gray.600';
      case 'blue':

      default:
        return 'blue.700';
    }
  }, [props.color]);

  return (
    <Box
      bg={color}
      borderTopLeftRadius="200"
      borderTopRightRadius="200"
      position="absolute"
      height="100"
      top="-1px"
      left="-1px"
      right="-1px"
    />
  );
}

Accent.displayName = 'Panel.Accent';

export default Accent;
