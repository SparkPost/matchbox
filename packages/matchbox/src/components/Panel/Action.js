import React from 'react';
import { Button } from '../Button';

const Action = React.forwardRef(function Action(props, userRef) {
  return <Button {...props} flat size="small" color="blue" ref={userRef} />;
});

Action.displayName = 'Panel.Action';
Action.propTypes = {
  ...Button.propTypes,
};
export default Action;
