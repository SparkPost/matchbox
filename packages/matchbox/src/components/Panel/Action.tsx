import React from 'react';
import { Button, PolymorphicButton } from '../Button';

const Action = React.forwardRef(function Action(props, userRef) {
  return <Button color="blue" {...props} variant="text" size="small" ref={userRef} />;
}) as PolymorphicButton;

Action.displayName = 'Panel.Action';

export default Action;
