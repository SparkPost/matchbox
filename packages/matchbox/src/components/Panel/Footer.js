import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Grid } from '../Grid';
import styles from './Panel.module.scss';

class Footer extends Component {

  static displayName = 'Panel.Footer';

  static propTypes = {
    /**
      * Left aligned content
      */
    left: PropTypes.node,

    /**
      * Right aligned content
      */
    right: PropTypes.node
  };

  render() {
    const { left, right, ...rest } = this.props;

    return (
      <div className={styles.Footer} {...rest}>
        <Grid>
          <Grid.Column xs={6}>
            <div className={styles.Left}>
              {left}
            </div>
          </Grid.Column>
          <Grid.Column xs={6}>
            <div className={styles.Right}>
              {right}
            </div>
          </Grid.Column>
        </Grid>
      </div>
    );
  }
}

export default Footer;
