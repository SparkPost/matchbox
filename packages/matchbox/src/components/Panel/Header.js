import React from 'react';
import { buttonsFrom } from '../Button';
import styles from './Panel.module.scss';

const actionOverrides = { flat: true, size: 'small' };

const Header = ({ title, actions }) => {
  const actionMarkup = actions && actions.length
    ? <div className={styles.Actions}>{buttonsFrom(actions, actionOverrides)}</div>
    : null;

  return (
    <div className={styles.Header}>
      <div className={styles.HeaderText}>
        {title}
      </div>
      {actionMarkup}
    </div>
  );
};

export default Header;
