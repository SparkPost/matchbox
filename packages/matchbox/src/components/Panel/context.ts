import React from 'react';

/**
 * Context is created here to pass system prop padding
 * to Panel.Header and Panel.Section
 */
export const PanelPaddingContext = React.createContext({});

/**
 * Context is created here to pass appearance prop
 * to Panel.Header, Panel.SubHeader, and Panel.Section
 */
export const PanelAppearanceContext = React.createContext({});
