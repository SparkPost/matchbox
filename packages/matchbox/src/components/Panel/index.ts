export { default as Panel } from './Panel';
export * from './Panel';
export * from './Section';
export * from './Header';
export * from './SubHeader';
export * from './Action';
