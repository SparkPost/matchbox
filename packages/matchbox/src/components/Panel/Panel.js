import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Section from './Section';
import Footer from './Footer';
import Header from './Header';
import styles from './Panel.module.scss';

class Panel extends Component {
  static Section = Section;
  static Footer = Footer;

  static displayName = 'Panel';

  static propTypes = {
    /**
     * The panel heading title
     */
    title: PropTypes.node,
    /**
      * Shows an accent bar, defaults to orange if boolean
      */
    accent: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.oneOf(['orange', 'blue', 'navy', 'purple', 'red', 'yellow', 'green', 'gray'])
    ]),
    /**
      * Adds a padded section automatically
      */
    sectioned: PropTypes.bool,
    /**
      * Actions that build buttons. Most button props will work in here.
      * e.g. { content: 'button label', onClick: callback() }
      */
    actions: PropTypes.arrayOf(PropTypes.shape({
      content: PropTypes.node.isRequired
    })),
    /**
      * Panel Content
      */
    children: PropTypes.node
  };

  render() {
    const {
      children,
      title,
      accent,
      sectioned,
      actions,
      className,
      ...rest
    } = this.props;

    const accentColor = accent === true ? 'orange' : accent;

    const headerMarkup = title
      ? <Header title={title} actions={actions} />
      : null;

    const contentMarkup = sectioned
      ? <Section>{children}</Section>
      : children;

    const accentMarkup = accentColor
      ? <div className={classnames(styles.Accent, styles[`accent-${accentColor}`])} />
      : null;

    const panelStyles = classnames(styles.Panel, className);

    return (
      <div className={panelStyles} {...rest}>
        {accentMarkup}
        {headerMarkup}
        {contentMarkup}
      </div>
    );
  }
}

export default Panel;
