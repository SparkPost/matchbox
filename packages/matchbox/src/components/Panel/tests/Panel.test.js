import React from 'react';
import Panel from '../Panel';
import { mount } from 'enzyme';

describe('Panel', () => {
  let wrapper;

  it('renders correctly', () => {
    wrapper = mount(
      <Panel accent sectioned title='title' actions={[{ content: 'Action' }]}>
        Foo
      </Panel>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with sections', () => {
    expect(mount(
      <Panel title='title'>
        <Panel.Section actions={[{ content: 'Action' }]}>Foo</Panel.Section>
        <Panel.Section>Bar</Panel.Section>
      </Panel>
    )).toMatchSnapshot();
  });

  it('renders correctly with footer', () => {
    expect(mount(
      <div>
        <Panel accent='red' />
        <Panel.Footer left={<span>left</span>} right='right' />
      </div>
    )).toMatchSnapshot();
  });
});
