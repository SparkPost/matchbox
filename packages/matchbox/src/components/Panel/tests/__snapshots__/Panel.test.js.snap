// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel renders correctly 1`] = `
<Panel
  accent={true}
  actions={
    Array [
      Object {
        "content": "Action",
      },
    ]
  }
  sectioned={true}
  title="title"
>
  <div
    className="Panel"
  >
    <div
      className="Accent accent-orange"
    />
    <Header
      actions={
        Array [
          Object {
            "content": "Action",
          },
        ]
      }
      title="title"
    >
      <div
        className="Header"
      >
        <div
          className="HeaderText"
        >
          title
        </div>
        <div
          className="Actions"
        >
          <Button.Group>
            <div
              className="Group"
            >
              <Button
                flat={true}
                key="0"
                size="small"
              >
                <button
                  className="Button flat small"
                  onMouseUp={[Function]}
                  type="button"
                >
                  Action
                </button>
              </Button>
            </div>
          </Button.Group>
        </div>
      </div>
    </Header>
    <Panel.Section>
      <div
        className="Body"
      >
        <div
          className="SectionContent"
        >
          Foo
        </div>
      </div>
    </Panel.Section>
  </div>
</Panel>
`;

exports[`Panel renders correctly with footer 1`] = `
<div>
  <Panel
    accent="red"
  >
    <div
      className="Panel"
    >
      <div
        className="Accent accent-red"
      />
    </div>
  </Panel>
  <Panel.Footer
    left={
      <span>
        left
      </span>
    }
    right="right"
  >
    <div
      className="Footer"
    >
      <Grid>
        <div
          className="Grid"
        >
          <Grid.Column
            xs={6}
          >
            <div
              className="Column xs-6"
            >
              <div
                className="Left"
              >
                <span>
                  left
                </span>
              </div>
            </div>
          </Grid.Column>
          <Grid.Column
            xs={6}
          >
            <div
              className="Column xs-6"
            >
              <div
                className="Right"
              >
                right
              </div>
            </div>
          </Grid.Column>
        </div>
      </Grid>
    </div>
  </Panel.Footer>
</div>
`;

exports[`Panel renders correctly with sections 1`] = `
<Panel
  title="title"
>
  <div
    className="Panel"
  >
    <Header
      title="title"
    >
      <div
        className="Header"
      >
        <div
          className="HeaderText"
        >
          title
        </div>
      </div>
    </Header>
    <Panel.Section
      actions={
        Array [
          Object {
            "content": "Action",
          },
        ]
      }
    >
      <div
        className="Body"
      >
        <div
          className="SectionContent"
        >
          Foo
        </div>
        <div
          className="Actions"
        >
          <Button.Group>
            <div
              className="Group"
            >
              <Button
                flat={true}
                key="0"
                size="small"
              >
                <button
                  className="Button flat small"
                  onMouseUp={[Function]}
                  type="button"
                >
                  Action
                </button>
              </Button>
            </div>
          </Button.Group>
        </div>
      </div>
    </Panel.Section>
    <Panel.Section>
      <div
        className="Body"
      >
        <div
          className="SectionContent"
        >
          Bar
        </div>
      </div>
    </Panel.Section>
  </div>
</Panel>
`;
